app/src/test/java/org/mozilla/fenix/components/metrics/AppStartupTypeTelemetryTest.kt

```diff
@@ -0,0 +1,118 @@
+
+import io.mockk.MockKAnnotations
+import io.mockk.impl.annotations.RelaxedMockK
+import io.mockk.verify
+import org.junit.Before
+import org.junit.Test
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.HOT
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.COLD_SAVED_INSTANCE
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.WARM_SAVED_INSTANCE
+
+class AppStartupTypeTelemetryTest {
+
+    @RelaxedMockK
+    private lateinit var metricController: MetricController
+
+    private lateinit var appStartupTypeTelemetry: AppStartupTypeTelemetry
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this)
+        appStartupTypeTelemetry = AppStartupTypeTelemetry(metricController)
+    }
+
+    @Test
+    fun `when FenixApplication and HomeActivity is created without savedInstance, then metric reports COLD startup type`() {
+        appStartupTypeTelemetry.onFenixApplicationOnCreate()
+        appStartupTypeTelemetry.onHomeActivityOnCreate(false)
+
+        val validType = COLD
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when FenixApplication and HomeActivity is created with savedInstance, then metric reports COLD_SAVED_INSTANCE startup type`() {
+        appStartupTypeTelemetry.onFenixApplicationOnCreate()
+        appStartupTypeTelemetry.onHomeActivityOnCreate(true)
+
+        val validType = COLD_SAVED_INSTANCE
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when application exists, and HomeActivity is created without savedInstanceState, then metric reports WARM startup type`() {
+        appStartupTypeTelemetry.onHomeActivityOnCreate(false)
+
+        val validType = WARM
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when application exists, and HomeActivity is created with savedInstanceState, then metric reports WARM_SAVED_INSTANCE startup type`() {
+        appStartupTypeTelemetry.onHomeActivityOnCreate(true)
+
+        val validType = WARM_SAVED_INSTANCE
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when application and HomeActivity exists, then metric reports HOT startup type`() {
+        appStartupTypeTelemetry.onHomeActivityOnRestart()
+
+        val validType = HOT
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when application comes to foreground and HomeActivity already exists, then metrics reports HOT startup type`() {
+        // application goes to background
+        appStartupTypeTelemetry.onApplicationOnStop()
```

Comment:
We're already in an impossible state: how can we call onStop without onCreate? Let's initialize the state of this object correctly (onCreate, onCreate then onStop). Here and below

Note also that this becomes a use-case for GIVEN (GIVEN the state WHEN onRestart...)